model/project_ref.go

```diff
@@ -530,6 +518,171 @@ func recursivelySetUndefinedFields(structToSet, structToDefaultFrom reflect.Valu
+	}
+	reflectedRepo := reflect.ValueOf(repoRef).Elem().Field(0) // specifically references the ProjectRef part of RepoRef
+	for i := 0; i < reflectedRepo.NumField(); i++ {
+		// for each field in the repo, look at each field in the project ref
+		var firstVal reflect.Value
+		allEqual := true
+		for j, pRef := range projectRefs {
+			reflectedBranchField := reflect.ValueOf(pRef).Field(i)
+			if j == 0 {
+				firstVal = reflectedBranchField
+			} else if !reflect.DeepEqual(firstVal.Interface(), reflectedBranchField.Interface()) {
+				allEqual = false
+				break
+			}
+		}
+		// if we got to the end of the loop, then all values are the same, so we can assign it to reflectedRepo
+		if allEqual {
+			reflectedRepo.Field(i).Set(firstVal)
+		}
+	}
+}
+
+func (p *ProjectRef) createNewRepoRef(u *user.DBUser) (repoRef *RepoRef, err error) {
+	repoRef = &RepoRef{ProjectRef{
+		Id:      mgobson.NewObjectId().Hex(),
+		Owner:   p.Owner,
+		Repo:    p.Repo,
+		Enabled: utility.TruePtr(),
+		Admins:  []string{},
+	}}
+
+	allEnabledProjects, err := FindMergedEnabledProjectRefsByOwnerAndRepo(p.Owner, p.Repo)
+	if err != nil {
+		return nil, errors.Wrap(err, "error finding all enabled projects")
+	}
+	// for every setting in the project ref, if all enabled projects have the same setting, then use that
+	defer func() {
+		err = recovery.HandlePanicWithError(recover(), err, "project and repo structures do not match")
+	}()
+	setRepoFieldsFromProjects(repoRef, allEnabledProjects)
+	repoRef.Admins = append(repoRef.Admins, u.Username())
+
+	// creates scope and give user admin access to repo
+	if err = repoRef.Add(u); err != nil {
+		return nil, errors.Wrapf(err, "problem adding new repo repo ref for '%s/%s'", p.Owner, p.Repo)
+	}
+	enabledProjectIds := []string{}
+	for _, p := range allEnabledProjects {
+		enabledProjectIds = append(enabledProjectIds, p.Id)
+	}
+	commonProjectVars, err := getCommonProjectVariables(enabledProjectIds)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting common project variables")
+	}
+	commonProjectVars.Id = repoRef.Id
+	if err = commonProjectVars.Insert(); err != nil {
+		return nil, errors.Wrap(err, "error inserting project variables for repo")
+	}
+
+	commonAliases, err := getCommonAliases(enabledProjectIds)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting common project aliases")
+	}
+	for _, a := range commonAliases {
+		a.ProjectID = repoRef.Id
+		if err = a.Upsert(); err != nil {
+			return nil, errors.Wrapf(err, "error upserting alias for repo")
+		}
+	}
+
+	return repoRef, nil
+}
+
+func getCommonAliases(projectIds []string) (ProjectAliases, error) {
+	commonAliases := map[string]ProjectAlias{} // use map for easy removal of non-common aliases
+	for i, id := range projectIds {
+		aliases, err := FindAliasesForProject(id)
+		if err != nil {
+			return nil, errors.Wrap(err, "error finding aliases for project")
+		}
+		if i == 0 {
+			for _, a := range aliases {
+				commonAliases[a.ID.Hex()] = a
```

Comment:
Since the ID is a uuid, does this make anything easier? I would think that this is the same as iterating through the array returned by FindAliasesForProject

Reply:
It doesn't work if you iterate forwards, because deleting an index will cause the for loop to skip something. If you loop backwards it should work